keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,437 @@
+- pick the largest one
+- turn off no more than 5 pods per minute
+
+### Implementation Details/Notes/Constraints
+
+#### Algorithm Pseudocode
+
+The algorithm to find the number of pods will look like this:
+
+```golang
+desiredReplicas = 0
+calculatedReplicas = AnyAlgorithmInHPAController(...)
+if calculatedReplicas > curReplicas:
+  constraint1 = CurReplicas * (1 + ScaleUpPercent/100)
+  constraint2 = CurReplicas + ScaleUpPods
+  scaleUpLimit = max(constraint1, constraint2)
+  desiredReplicas = min(scaleUpLimit, calculatedReplicas)
+else if calculatedReplicas < CurReplicas:
+  constraint1 = curReplicas * (1 - ScaleDownPercent/100)
+  constraint2 = CurReplicas - ScaleDownPods
+  scaleDownLimit = max(constraint1, constraint2)
+  desiredReplicas = max(scaleDownLimit, calculatedReplicas)
+```
+
+I.e., from the two provided constraints the larger one is used.
+
+If you don’t want to scale, you should set both parameters to zero for the appropriate direction (Up/Down).
+
+If only one parameter is given and the other is 0, then use the only defined constraint.
+
+The algorithm for `delay` will look like this:
+
+```golang
+for { // infinite cycle inside the HPA controller
+  desiredReplicas = calculateDesiredReplicas()
+  if desiredReplicas > curReplicas:
+    storeRecommendation(desiredReplicas, scaleUpRecommendations)
+    recommendations = getLastRecommendations(scaleUpRecommendations, ScaleUpDelaySeconds) // get recommendations for the last ScaleUpDelaySeconds
+    nextReplicas = min(recommendtaions)
+  if desiredReplicas < curReplicas:
+    storeRecommendation(desiredReplicas, scaleDownRecommendations)
+    recommendations = getLastRecommendations(scaleDownRecommendations, ScaleDownDelaySeconds) // get recommendations for the last ScaleDownDelaySeconds
+    nextReplicas = max(recommendtaions)
+  setReplicas(nextReplicas)
+  sleep(ControllerSleepTime)
+}
+```
+
+Effectively, this is a full copy of the current [Stabilization Window][] algorithm:
+
+- While scaling up, we should pick the safest (smallest) "desiredReplicas" number during last `delaySeconds`.
+- While scaling down, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
+
+The “stabilization window" as a result becomes an alias for the `constraints.scaleDown.delaySeconds`.
+
+
+[Stabilization Window]: #stabilization-window
+
+#### Default Values
+
+For smooth transition it makes sense to set the following default values:
+
+- `constraints.scaleUp.delaySeconds = 0`, the delay is not used, instantly scale up
+- `constraints.scaleDown.delaySeconds = 300`, wait 5 min for the largest recommendation and then scale down to that value.
+- `constraints.scaleUp.rate.periodSeconds = 60`, one minute period for scaleUp
+- `constraints.scaleDown.rate.periodSeconds = 60`, one minute period for scaleDown
+- `constraints.scaleUp.rate.percent = 100`, allow to twice the number of pods
+- `constraints.scaleUp.rate.pods = 4`, i.e. allow adding up to 4 pods
```

Comment:
I don't think the current HPA has a default scale limit of 4 pods.  I think the default should be no pod limit.

Reply:
It is 4 nevertheless:
https://github.com/kubernetes/kubernetes/blob/release-1.14/pkg/controller/podautoscaler/horizontal.go#L56

Reply:
That `ScaleUpMinimum` has different semantics than `constraints.scaleUp.rate.pods`.  It is used in combination with scale up limit rate to let it get off the ground (`max(replicas,ScaleUpMinimum)`).  You can avoid the need for this particular parameter by storing recommendations as float64 so you can "accumulate" scale to achieve whole pods.

The `contraints.scaleUp.rate.pods` is a separate constraint which establishes the scale-up budget in whole pods over a duration.  There is no such limit currently, so it should default to `nil`.

Reply:
(chatting with @gliush about this)

I think I was wrong here.  Because we apply the `max` of rate constraints, the `ScaleUpMinimum` and `constraints.scaleUp.rate.pods` do have the same semantics.  Defaulting to `4` is correct.

Disregard.  :)

Reply:
:+1: